modules/function/src/main/java/com/opengamma/strata/function/calculation/swaption/DefaultSwaptionMarketLookup.java

```diff
@@ -0,0 +1,371 @@
+import org.joda.beans.impl.direct.DirectMetaProperty;
+import org.joda.beans.impl.direct.DirectMetaPropertyMap;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.opengamma.strata.basics.index.IborIndex;
+import com.opengamma.strata.basics.index.Index;
+import com.opengamma.strata.basics.market.MarketData;
+import com.opengamma.strata.basics.market.MarketDataKey;
+import com.opengamma.strata.calc.CalculationRules;
+import com.opengamma.strata.calc.marketdata.CalculationEnvironment;
+import com.opengamma.strata.calc.marketdata.FunctionRequirements;
+import com.opengamma.strata.calc.runner.CalculationParameter;
+import com.opengamma.strata.collect.Messages;
+import com.opengamma.strata.market.id.SwaptionVolatilitiesId;
+import com.opengamma.strata.market.view.SwaptionVolatilities;
+
+/**
+ * The swaption lookup, used to select volatilities for pricing.
+ * <p>
+ * This provides swaption volatilities by index.
+ * <p>
+ * The lookup implements {@link CalculationParameter} and is used by passing it
+ * as an argument to {@link CalculationRules}. It provides the link between the
+ * data that the function needs and the data that is available in {@link CalculationEnvironment}.
+ */
+@BeanDefinition(builderScope = "private")
+final class DefaultSwaptionMarketLookup
+    implements SwaptionMarketLookup, ImmutableBean, Serializable {
+
+  /**
+   * The volatility identifiers, keyed by index.
+   */
+  @PropertyDefinition(validate = "notNull")
+  private final ImmutableMap<IborIndex, SwaptionVolatilitiesId> volatilityIds;
+
+  //-------------------------------------------------------------------------
+  /**
+   * Obtains an instance based on a single mapping from index to volatility identifier.
+   * <p>
+   * The lookup provides volatilities for the specified index.
+   *
+   * @param index  the Ibor index
+   * @param volatilityId  the volatility identifier
+   * @return the swaption lookup containing the specified mapping
+   */
+  public static DefaultSwaptionMarketLookup of(IborIndex index, SwaptionVolatilitiesId volatilityId) {
+    return new DefaultSwaptionMarketLookup(ImmutableMap.of(index, volatilityId));
+  }
+
+  /**
+   * Obtains an instance based on a map of volatility identifiers.
+   * <p>
+   * The map is used to specify the appropriate volatilities to use for each index.
+   *
+   * @param volatilityIds  the volatility identifiers, keyed by index
+   * @return the swaption lookup containing the specified volatilities
+   */
+  public static DefaultSwaptionMarketLookup of(Map<IborIndex, SwaptionVolatilitiesId> volatilityIds) {
+    return new DefaultSwaptionMarketLookup(volatilityIds);
+  }
+
+  //-------------------------------------------------------------------------
+  @Override
+  public ImmutableSet<IborIndex> getVolatilityIndices() {
+    return volatilityIds.keySet();
+  }
+
+  @Override
+  public ImmutableSet<MarketDataKey<?>> getVolatilityIds(IborIndex index) {
```

Comment:
The lookup interface is at a higher level than the market data interfaces. Where market data is opaque, and cannot tell you what it contains, the lookup objects can tell you what they contain. (A lookup implementation is typically going to be a hashmap from some kind of key to some kind of market data identifier). The `RatesMarketView` makes use of this when doing bucketed gamma. These swaption methods parallel those on rates.